services/paymaster/src/api.rs

```diff
@@ -51,73 +55,186 @@ pub async fn validate_transaction(
         ));
     }
 
-    transaction
-        .message
-        .instructions()
-        .iter()
-        .try_for_each(|instruction| {
-            let program_id = instruction.program_id(transaction.message.static_account_keys());
-            if !program_whitelist.contains(program_id) {
-                return Err((
+    let instructions = transaction.message.instructions();
+    instructions.iter().try_for_each(|instruction| {
+        let program_id = instruction.program_id(transaction.message.static_account_keys());
+        if !program_whitelist.contains(program_id) {
+            return Err((
+                StatusCode::BAD_REQUEST,
+                format!("Transaction contains unauthorized program ID: {program_id}"),
+            ));
+        }
+        Ok(())
+    })?;
+
+    // If the transaction is using the tollbooth, we don't need to simulate it
+    if instructions
+        .first()
+        .map(|i| i.program_id(transaction.message.static_account_keys()) == &TOLLBOOTH_PROGRAM_ID)
+        .unwrap_or_default()
```

Comment:
The code in this conditional I think can be cleaned up so you don't have to make the expectations and can more clearly enumerate the cases.  Something like this:

```rs
let first_instruction_program_id = instruction.first().map(|i| i.program_id(transaction.message.static_account_keys());
let last_instruction_program_id = instructions.last().map(|i|i.program_id(transaction.message.static_account_keys());

match (first_instruction_program_id, last_instruction_program_id) {
  (Some(&TOLLBOOTH_PROGRAM_ID), Some(&TOLLBOOTH_PROGRAM_ID)) => ...,
  ...
}
```